### PR TITLE
[Clang] Fix the order of addInstantiatedParameters in LambdaScopeForCallOperatorInstantiationRAII

### PR DESCRIPTION
**Pull request by LYP951018**
Currently, `addInstantiatedParameters` is called from the innermost lambda outward. However, when the function parameters of an inner lambda depend on the function parameters of an outer lambda, it can lead to a crash due to the inability to find a mapping for the instantiated decl.

This PR corrects this behavior by calling `addInstantiatedParameters` from the outside in.

repro code: https://godbolt.org/z/KbsxWesW6

```cpp
namespace dependent_param_concept {
template <typename... Ts> void sink(Ts...) {}
void dependent_param() {
  auto L = [](auto... x) {
    return [](decltype(x)... y) { // `y` depends on `x`
      return [](int z)
        requires requires { sink(y..., z); }
      {};
    };
  };
  L(0, 1)(1, 2)(1);
}
} // namespace dependent_param_concept
```

This PR is a prerequisite for implmenting #61426

### CODE REVIEW

**Comment by LYP951018** [comment]
Sorry for disturbing... I'm trying to get the CI working. I can't figure out why buildkite/github-pull-requests failed...

File: clang/lib/Sema/SemaLambda.cpp [hunk 1]
```diff
@@ -2379,23 +2379,32 @@ Sema::LambdaScopeForCallOperatorInstantiationRAII::
 
   SemaRef.RebuildLambdaScopeInfo(cast<CXXMethodDecl>(FD));
 
-  FunctionDecl *Pattern = getPatternFunctionDecl(FD);
-  if (Pattern) {
-    SemaRef.addInstantiatedCapturesToScope(FD, Pattern, Scope, MLTAL);
+  FunctionDecl *FDPattern = getPatternFunctionDecl(FD);
+  if (!FDPattern)
+    return;
```
Comment (zyn0217) [inline clang/lib/Sema/SemaLambda.cpp:2385]: nit: Can you revert unrelated changes?

File: clang/lib/Sema/SemaLambda.cpp [hunk 2]
```diff
@@ -2379,23 +2379,32 @@ Sema::LambdaScopeForCallOperatorInstantiationRAII::
...
-      if (!ParentFD || !Pattern)
-        break;
+  llvm::SmallVector<std::pair<FunctionDecl *, FunctionDecl *>, 4>
+      ParentInstantiations;
+  std::pair<FunctionDecl *, FunctionDecl *> Current = {FDPattern, FD};
+  while (true) {
+    Current.first = dyn_cast<FunctionDecl>(
+        getLambdaAwareParentOfDeclContext(Current.first));
+    Current.second = dyn_cast<FunctionDecl>(
+        getLambdaAwareParentOfDeclContext(Current.second));
```
Comment (zyn0217) [inline clang/lib/Sema/SemaLambda.cpp:2398]: Using `.first` and `.second` seems a bit unclear. Can we wrap them into a local struct?

File: clang/lib/Sema/SemaLambda.cpp [hunk 3]
```diff
@@ -2379,23 +2379,32 @@ Sema::LambdaScopeForCallOperatorInstantiationRAII::
...
-      SemaRef.addInstantiatedParametersToScope(ParentFD, Pattern, Scope, MLTAL);
-      SemaRef.addInstantiatedLocalVarsToScope(ParentFD, Pattern, Scope);
-    }
+    if (!Current.first || !Current.second)
+      break;
+
+    ParentInstantiations.push_back(Current);
+  }
+
+  for (const auto &[Pattern, Inst] : llvm::reverse(ParentInstantiations)) {
```
Comment (zyn0217) [inline clang/lib/Sema/SemaLambda.cpp:2406]: I think it would be better to retain the name `PatternFD`.
In addition, can you also add some comments to explain why we're enumerating instantiations in the reversed way here? An example would be
```
// Add instantiated parameters to scopes in the order from the outer lambda to
// the inner lambda. We do so because the inner parameters could depend on
// any outer template parameters.
```

**Review by zyn0217 — COMMENTED** [review]
Thanks! I think the approach makes sense. Just some nits from me
@cor3ntin do you have any other comments? I think you can do the last approval. :)

**Comment by zyn0217** [comment]
Oh, I forgot it: this also needs a release note.

File: clang/lib/Sema/SemaLambda.cpp (hunk 2 above)
Comment (cor3ntin) [reply clang/lib/Sema/SemaLambda.cpp:2398]: Alternatively, only form the pair when emplacing in the vector.

File: clang/lib/Sema/SemaLambda.cpp (hunk 1 above)
Comment (LYP951018) [reply clang/lib/Sema/SemaLambda.cpp:2385]: I would prefer to keep the current change. Firstly, this change is minimal and won’t significantly impact the review workload for this PR. Secondly, it reduces one level of indentation, which slightly improves readability. ;)

**Review by mizvekov — APPROVED** [review]
LGTM

**Comment by LYP951018** [comment]
Thanks for the review ;)
